File: built-in plugins dsp/mb-limiter-soft.cpp
/**
 * MB Soft Limiter
 * Category : effect
 * Type     : limiter
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Gentle peak limiting
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_LIMITER_SOFT_H
#define MB_LIMITER_SOFT_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbLimiterSoft : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-limiter-soft";
    static constexpr const char* PLUGIN_NAME    = "MB Soft Limiter";
    static constexpr const char* PLUGIN_TYPE    = "limiter";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float ceiling = -1f;  // range [-12, 0]
    float knee = 6f;  // range [0, 12]
    };

    MbLimiterSoft() = default;
    ~MbLimiterSoft() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.ceiling = std::clamp(params.ceiling, -12f, 0f);
        params.knee = std::clamp(params.knee, 0f, 12f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Soft Limiter
        return input;
    }
};

#endif // MB_LIMITER_SOFT_H

File: maxbooster7.5/built-in plugins dsp/mb-limiter-soft.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbLimiterSoftPlugin: PluginDefinition = {
  id: "mb-limiter-soft",
  slug: "mb-limiter-soft",
  name: "MB Soft Limiter",
  category: "effect",
  type: "limiter",
  version: "1.0.0",
  description: "Gentle peak limiting",
  author: "Max Booster",
  grade: "A",
  parameters: [
    {
      id: "ceiling",
      name: "Ceiling",
      type: "float",
      defaultValue: -1,
      minValue: -12,
      maxValue: 0,
      automatable: true,
    },
    {
      id: "knee",
      name: "Knee",
      type: "float",
      defaultValue: 6,
      minValue: 0,
      maxValue: 12,
      automatable: true,
    },
  ],
  defaultPreset: { ceiling: -1, knee: 6 },
};

export default MbLimiterSoftPlugin;
